File: kotlin/day4_part2_2017.kt
import java.io.File

fun main() {
    val passphrases = File("input.txt").readLines()

    val validPassphrasesPart1 = passphrases.filter { isValidPassphrasePart1(it) }.size
    val validPassphrasesPart2 = passphrases.filter { isValidPassphrasePart2(it) }.size

    println("Part 1: $validPassphrasesPart1")
    println("Part 2: $validPassphrasesPart2")
}

fun isValidPassphrasePart1(passphrase: String): Boolean {
    val words = passphrase.split(" ")
    return words.toSet().size == words.size
}

fun isValidPassphrasePart2(passphrase: String): Boolean {
    val words = passphrase.split(" ")
    val sortedWords = words.map { it.toCharArray().sorted().joinToString("") }
    return sortedWords.toSet().size == sortedWords.size
}

File: clojure/day4_part2_2017.clj

(defn valid-passphrase [passphrase]
  (let [words (clojure.string/split passphrase #" ")
        sorted-words (map #(apply str (sort %)) words)]
    (= (count words) (count (distinct words)))))

(defn valid-passphrase-part-two [passphrase]
  (let [words (clojure.string/split passphrase #" ")
        sorted-words (map #(apply str (sort %)) words)]
    (= (count words) (count (distinct sorted-words)))))

(defn count-valid-passphrases [passphrases]
  (count (filter valid-passphrase passphrases)))

(defn count-valid-passphrases-part-two [passphrases]
  (count (filter valid-passphrase-part-two passphrases)))

(def input (slurp "input.txt"))
(def passphrases (clojure.string/split-lines input))

(println (count-valid-passphrases passphrases))
(println (count-valid-passphrases-part-two passphrases))
